File: eratos.php
<?php
function sieve(int $limit): array {
    $sieve = array_fill(0, $limit + 1, true);
    $sieve[0] = $sieve[1] = false;

    for ($i = 2; $i * $i <= $limit; $i++) {
        if ($sieve[$i]) {
            for ($j = $i * $i; $j <= $limit; $j += $i) {
                $sieve[$j] = false;
            }
        }
    }

    $primes = [];
    foreach ($sieve as $num => $isPrime) {
        if ($isPrime) $primes[] = $num;
    }
    return $primes;
}

// -----------------------------

$limit = isset($_GET['limit']) ? (int)$_GET['limit'] : 500_000;
if ($limit > 50_000_000) $limit = 50_000_000;

$start = hrtime(true);
$primes = sieve($limit);
$end = hrtime(true);

$count = count($primes);
$elapsed = number_format(($end - $start) / 1e9, 6);

echo "PHP found $count primes up to $limit in {$elapsed} seconds.";

File: src/bin/eratos.rs
use std::time::Instant;

fn sieve(limit: usize) -> Vec<usize> {
    let mut sieve = vec![true; limit + 1];
    sieve[0] = false;
    sieve[1] = false;

    let mut p = 2;
    while p * p <= limit {
        if sieve[p] {
            let mut j = p * p;
            while j <= limit {
                sieve[j] = false;
                j += p;
            }
        }
        p += 1;
    }

    sieve
        .iter()
        .enumerate()
        .filter_map(|(i, &is_prime)| if is_prime { Some(i) } else { None })
        .collect()
}

fn main() {
    let limit: usize = std::env::args()
        .nth(1)
        .and_then(|x| x.parse().ok())
        .unwrap_or(500_000);

    let limit = limit.min(500_000_000);

    let start = Instant::now();
    let primes = sieve(limit);
    let elapsed = start.elapsed().as_secs_f64();

    println!(
        "Rust found {} primes up to {} in {:.6} seconds.",
        primes.len(),
        limit,
        elapsed
    );
}
